core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java

```diff
@@ -0,0 +1,438 @@
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * A regrettable necessity. The Kotlin SDK relies on an outdated Ktor HTTP client based on Kotlin
+   * Coroutines which are difficult to work with in Java. Here we make a HTTP client to handle the
+   * request and input stream ourselves. This adds complexity that would normally be handled by the
+   * SDK. We should revisit this once the SDK has built-in streams.
+   *
+   * TODO https://github.com/looker-open-source/sdk-codegen/issues/1341:
+   *  Add streaming support to the Kotlin SDK.
+   */
+  protected InputStream makeRunQueryRequest(String url) throws IOException {
+    AuthSession authSession = getSdk().getAuthSession();
+    Transport sdkTransport = authSession.getTransport();
+
+    // makes a proper URL from the API endpoint path as the SDK would.
+    String endpoint = sdkTransport.makeUrl(url, Collections.emptyMap(), null);
+    URL httpsUrl = new URL(endpoint);
+    HttpsURLConnection connection = (HttpsURLConnection) httpsUrl.openConnection();
+
+    // WARNING: You should only set `verifySSL=false` for local/dev instances!!
+    if (!sdkTransport.getOptions().getVerifySSL()) {
+      trustAllHosts(connection);
+    }
+
+    // timeout is given as seconds
+    int timeout = sdkTransport.getOptions().getTimeout() * 1000;
+    connection.setReadTimeout(timeout);
+    connection.setRequestMethod("GET");
+    connection.setRequestProperty("Accept", "application/json");
+    connection.setDoOutput(true);
+
+    // Set the auth header as the SDK would
+    connection.setRequestProperty("Authorization",
+        "token " + authSession.getAuthToken().getAccessToken());
+
+    int responseCode = connection.getResponseCode();
+    if (responseCode == 200) {
+      // return the input stream to parse from.
+      return connection.getInputStream();
+    } else {
+      throw new IOException("HTTP request failed with status code: " + responseCode);
+    }
+  }
+
+  private void seekToRows(JsonParser parser) throws IOException {
+    while (parser.nextToken() != null && !ROWS_KEY.equals(parser.currentName())) {
+      // move position to start of `rows`
+    }
+  }
+
+  private void seekToValue(JsonParser parser) throws IOException {
+    while (parser.nextToken() != null && !VALUE_KEY.equals(parser.currentName())) {
+      // seeking to `value` key for the field e.g. `"rows": [{"field_1": {"value": 123 }}]
+    }
+    // now move to the actual value
+    parser.nextToken();
+  }
+
+  private void putExceptionOrFail(BlockingQueue queue, Exception e) {
+    try {
+      // `put` blocks until there is room on the queue but needs a catch
+      queue.put(makeException(e));
+    } catch (InterruptedException ex) {
+      throw new RuntimeException(ex);
+    }
+  }
+
+  /**
+   * Prepares a thread to stream a query response into a series of {@link FrameEnvelope}s.
+   */
+  protected Thread prepareStreamingThread(String baseUrl, Signature signature, int fetchSize,
```

Comment:
How about breaking this and some of the associated parsing logic into a separate class, and leaving `LookerRemoteMeta` as a thing which spawns threads and connects them to queues? The class is a bit big now and it's not clear which code runs on the caller's thread and which code runs in a streaming thread.

The stream parser itself probably doesn't need to know about threads, and if it takes a `Consumer<FrameEnvelope>` argument it needn't know about queues either. The value of the consumer argument could be a lambda wrapping `blockingQueue::put` that handles `InterruptedException`.

Then if someone wants to test the parser in isolation, they don't need to mess around with threads or queues.

Reply:
Moved into `LookerResponseParser` class and added some tests.